File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-16.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
       vector<int> chars(128,0);
        int res = 0;
        int left , right = 0;

        while(right<n){
            char ch = s[right];
            if(chars[ch]==0){
                chars[ch]++;
                res = max(res,right-left+1);
                right++;
            }
            else {
                chars[s[left]]--;
                left++;
            }
        }
        return res;

        
    }
};

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-16.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        count = {}      # plain dict, no keys
        res = 0
        left = 0
        maxf = 0


        for right in range(len(s)):
            count[s[right]]=1+count.get(s[right],0)   # .get supplies 0 if missing
            maxf = max(maxf,count[s[right]])          # most frequent char in window

            while(right-left+1)-maxf > k:             # invalid: replacements > k
                count[s[left]]-=1
                left+=1
            res = max(res,right-left+1)               # track longest valid window

        return res